Source/ARTHttp.m

```diff
@@ -47,14 +47,22 @@ - (void)dealloc {
 }
 
 - (NSObject<ARTCancellable> *)executeRequest:(NSMutableURLRequest *)request completion:(void (^)(NSHTTPURLResponse *_Nullable, NSData *_Nullable, NSError *_Nullable))callback {
-    [self.logger debug:@"--> %@ %@\n  Body: %@\n  Headers: %@", request.HTTPMethod, request.URL.absoluteString, [[NSString alloc] initWithData:request.HTTPBody encoding:NSUTF8StringEncoding], request.allHTTPHeaderFields];
+    NSString *requestBodyStr = [[NSString alloc] initWithData:request.HTTPBody encoding:NSUTF8StringEncoding];
+    if (!requestBodyStr) {
+        requestBodyStr = [request.HTTPBody base64EncodedStringWithOptions:NSDataBase64EncodingEndLineWithCarriageReturn];
+    }
+    [self.logger debug:@"--> %@ %@\n  Body: %@\n  Headers: %@", request.HTTPMethod, request.URL.absoluteString, requestBodyStr, request.allHTTPHeaderFields];
```

Comment:
This new code adds runtime overhead that's required for normal function. Do we have the capability to detect at runtime if debug logs are enabled and, therefore, only construct this string in that case?

Also, why can the call to `initWithData:encoding:` fail? And if it can fail, would it not be better to encapsulate this logic (used twice here) in to a function to perhaps document why it can fail in the one place?

Reply:
@QuintinWillison I agree, it's necessary overhead for the normal use case. I'll find out a solution. `initWithData:encoding:` can fail when the data is encoded with MsgPack and that's why it should at least show a B64 string instead of null. When I was debugging I thought: `Oh, here's the problem, the body is empty` and that was not the case 😄